aten/src/ATen/core/dynamic_type.h

```diff
@@ -0,0 +1,118 @@
+  _(Bool, DYNAMIC_TYPE_BIT(2))                                               \
+  _(Int, kDynamicIntTypeBit)                                                 \
+  _(Float, kDynamicFloatTypeBit)                                             \
+  _(Complex, kDynamicComplexTypeBit)                                         \
+  _(Number,                                                                  \
+    (kDynamicIntTypeBit | kDynamicFloatTypeBit | kDynamicComplexTypeBit))    \
+  _(String, DYNAMIC_TYPE_BIT(6))                                             \
+  _(List, kDynamicListTypeBit)                                               \
+  _(Tuple, (kDynamicTupleTypeBit | kDynamicCovariantTypeBit))                \
+  _(Dict, DYNAMIC_TYPE_BIT(9))                                               \
+  _(Class, DYNAMIC_TYPE_BIT(10))                                             \
+  _(Optional,                                                                \
+    (DYNAMIC_TYPE_BIT(11) | kDynamicNoneTypeBit | kDynamicCovariantTypeBit)) \
+  _(AnyList, (kDynamicListTypeBit | kDynamicAnyTypeBit))                     \
+  _(AnyTuple,                                                                \
+    (kDynamicTupleTypeBit | kDynamicCovariantTypeBit | kDynamicAnyTypeBit))  \
+  _(Any, 0xffffffff)
+
+namespace c10 {
+
+struct FunctionSchema;
+struct LabeledDynamicType;
+class DynamicType;
+using DynamicTypePtr = std::shared_ptr<DynamicType>;
+
+// Low dependency jit type with minimal subtyping and structucing support,
+// designed for embedded cases.
+class DynamicType : public Type {
+  using ClassTypePtr = std::shared_ptr<const c10::ClassType>;
+
+ public:
+  ~DynamicType() override;
+
+  struct Arguments {
+    Arguments() = default;
+    Arguments(c10::ArrayRef<TypePtr>);
+    Arguments(const c10::FunctionSchema&, c10::ArrayRef<TypePtr>);
+    std::vector<LabeledDynamicType> elems;
+  };
+
+  enum class Tag : DynamicTypeBits {
+#define DYNAMIC_TYPE_ITEM(NAME, VAL) NAME = VAL,
+    FORALL_DYNAMIC_TYPES(DYNAMIC_TYPE_ITEM)
+#undef DYNAMIC_TYPE_ITEM
+        Singleton = kDynamicSingletonTypeBit,
+  };
+
+  bool operator==(const Type& rhs) const override;
+  bool isSubtypeOfExt(const Type& rhs, std::ostream* why_not) const override;
+  std::string str() const override {
+    return "Dynamic";
+  }
+  static const TypeKind Kind = TypeKind::DynamicType;
+  static DynamicTypePtr create(Type& ty);
+
+ private:
+  friend struct Type;
+  static std::shared_ptr<const DynamicType> create(const Type& ty);
+  DynamicType(const Type& other);
+  bool equals(const DynamicType& other) const;
+
+  template <typename F>
+  bool compare(const DynamicType& other, F&& f) const {
+    if (arguments_.elems.size() != other.arguments_.elems.size()) {
+      return false;
+    }
+    for (size_t i = 0; i < arguments_.elems.size(); i++) {
+      if (!f(arguments_.elems[i], other.arguments_.elems[i])) {
+        return false;
+      }
+    }
+    return true;
+  }
```

Comment:
Is this implemented mainly to save code duplication?

It is slightly odd that `compare` bakes in part of comparison logic (like number of arguments must match) but allowing customization of how each pair of arguments are compared.

Reply:
And in call sites of `compare`, looks like type tag is always compared separately beforehand. So `compare` is more like `compare_arguments`?

Reply:
yeah, for code duplication concerns.
I should name it `compareArguments` probably. The purpose is to have an algorithm for comparing arguments, used in subtype and equality tests.